# Determine the learning rate

The learning rate $alpha$ is different according to optimization target.

## Min Potential Problem

Assume a point distribution evenly enough, according to shell theorem, the force excert on any point would be

$$
F = \lim_{
  n\to +\infty
}{
  n-1
}
$$

Note this is the force towards the radial direction.

Assume hexagon distribution of edge length $a$, then the average distance is $d=\sqrt{3}a$, while the area of such a hexagon would be

$$
A = \lim_{
  n\to +\infty
}{
  6 \times \cfrac{
    \sqrt{3}a^2
  }{
    4
  }
} = \cfrac{
  4\pi
}{
  3n
}
$$

The position difference in the circular dirction wihin iteration should be much less than the length, thus the safe upperbound of learning rate $\alpha$ is

$$
\alpha = \lim_{
  n\to +\infty
}{
  \cfrac{
    d
  }{
    F
  }
} = \cfrac{
  2\sqrt{2\sqrt{3}\pi}
}{
  3
} \cfrac{1}{\sqrt{n}} \cfrac{1}{F}
$$

In [1]:
import numpy as np
2 * np.sqrt(2*np.sqrt(3)*np.pi)/3

2.1992722215825355

However, in early steps, F might be far larger than $n-1$. This calls for a dynamic alpha decided by the largest F.

$$
\alpha = \cfrac{
  2.199
}{
  \sqrt{n}F
}
$$

Note this is an upperbound, the actual learning rate should be less than this to avoid fluctuating around the minimum.